Implement restaurant lookup

naraekn opened the pull request:
아직 많이 하지는 못했지만, 주간 회고에 대하여 피드백을 받고ㅎㅎ pull request를 우선 보냅니다
집 도착하고 더 코딩해서 푸시하겠습니다 😄

저번주에 과제를 했을 때보다 TDD의 단위를 작게하려고 노력해봤어요!

naraekn committed:
Add test for App

Test if categories are rendered

naraekn committed:
Implement App

Implement App with mock data

naraekn committed:
Refactor App

Use buttons instead of paragraph
Use categories object lists

naraekn committed:
Add provider to index

naraekn committed:
Add store

naraekn committed:
Add test for reducer

Test if reducer returns initial state with example data

naraekn committed:
Implement reducer

Use initialState with example data
Return initialState for all cases

naraekn committed:
Update test for app to add useSelector

naraekn committed:
Add react-redux manual mocks

naraekn committed:
Add test data in fixture

naraekn committed:
Update App to add useSelector

Relocate useSelector to fix errors in App test

naraekn committed:
Add useEffect to set initial categories

Add dispatch
Add loadCategories

naraekn committed:
Add setCategories action

naraekn committed:
Update test for reducer in case of setCategories action

naraekn committed:
Update reducer to cover setCategories action

hannut91 commented on src/App.test.jsx:
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import { useSelector } from 'react-redux';
```
위에 빈 줄 추가해서 리액트라이브러리와 리덕스 라이브러리를 구분해주면 좋을 것 같아요

naraekn replied:
빈줄 추가에 대한 피드백을 많이 받았는데, 궁금한점이 있어서 여쭤봅니다.

```import``` 를 할 때 저는,
1) @ 없는 노드모듈
2) @ 있는 노드모듈
3) src 폴더 내 파일
4) src 폴더 외 파일

위 세 개를 차례대로 썼어요. 각 그룹이 비슷한 애들일 거라고 생각해서(?), 1~4번 그룹 내에서는 띄어쓰기를 안하고 각 그룹 사이에만 빈줄 추가를 넣으려고 했던 거 같아요. (적고 보니 좋은 기준은 아니었던 거 같네요ㅎㅎㅎ 그냥 생각 없이 저만의 기준을 적용하고 있었다는 점을 오늘 깨달았습니다.) 

오늘뿐만 아니라 이전에도 몇 번 '빈줄 추가'에 대한 피드백을 받았는데, 돌아보니 특별히 생각하지 않고 피드백만 반영해오다보니 같은 지적을 지속적으로 받은거 같다는 생각이 들었어요! '빈줄 추가'의 좋은 기준이 무엇인지에 대하여 고민을 하게 되네요.

모든 코드에서 '빈 줄 추가'에 대한 완벽한 기준은 없을 거라고 생각하지만, 멘토님의 의견이 궁금합니다.

* ```import```문 사이에서 빈줄을 추가하는 기준이 무엇일까요?
* ```import``` 뿐만 아니라 모든 코드에서 빈 줄을 언제 추가하는게 좋을까요?

hannut91 replied:
`import`문 사이에서 빈줄을 추가하는 기준은 서로 관련이 있는 것들끼리 묶는거에요. `React`라이브러리와 `Redux`라이브러리는 서로 다르니까 빈 줄을 추가하는게 맞다고 생각했습니다. 이 또한 팀마다 다를 수 있으므로 컨벤션에 따라가는게 사실 가장 정답입니다.

모든 코드에서 빈 줄을 추가하는 것도 관련이 있는 것들끼리 묶는거에요. 관련이 없는 것들을 일부러 빈 줄을 추가해서 구분해줄 수 있습니다.

naraekn replied:
ㅎㅎ 질문 해놓고 너무 이상한 곳에 꽂힌건가 생각했는데, 답변해주셔서 감사해요 :)

음.. 의미 단위로 나누는게 좋은거군요! 저는 '형식' 단위로 빈줄을 추가하는 경우가 많았던 거 같네요
앞으로는 한번 의미 단위로 생각해봐야겠습니다!

hannut91 commented on src/App.test.jsx:
```diff
@@ -0,0 +1,22 @@
+import { render } from '@testing-library/react';
+
+import App from './App';
+import { categories } from '../fixture/test-data';
```
위에 빈 줄 추가해서 구분해주면 좋을 것 같아요

hannut91 commented on src/index.jsx:
```diff
@@ -1,9 +1,13 @@
 import React from 'react';
 import ReactDOM from 'react-dom';
+import { Provider } from 'react-redux';
```
위에 빈 줄 하나를 추가해서 리액트 라이브러리 임포트와 리덕스 라이브러리 임포트를 구분해주면 좋을 것 같아요

hannut91 commented on src/index.jsx:
```diff
@@ -1,9 +1,13 @@
+import { Provider } from 'react-redux';
 
 import App from './App';
+import store from './store';
```
위에 빈 줄 하나를 추가해서 구분해주면 좋을 것 같아요

hannut91 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,26 @@
+import { setCategories } from './actions';
+
+describe('reducer', () => {
+  const testCategories = [
+    { id: 1, name: '한식' },
+    { id: 2, name: '중식' },
+  ];
```
이것도 fixture를 사용하면 좋을 것 같아요

hannut91 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,26 @@
+    { id: 2, name: '중식' },
+  ];
+
+  context('when reducer is executed first time', () => {
+    it('returns initial state', () => {
+      const state = reducer(undefined, 'testAction');
+
+      expect(state.categories).toHaveLength(0);
+    });
+  });
```
테스트 대상이 정확히 무엇인지 파악하기 어려운 것 같아요. 잘못된 액션이 들어왔을 때를 확인하고 싶으셨던건가요?

naraekn replied:
처음에 ```reducer```에 ```state```이 없을 때, ```initialState```이 제대로 반환되는지 확인하고 싶었어요! 제가 처음에 ```reducer.test```를 구현했을 때는 정의해놓은 액션이 없는 상태여서 우선 아무 액션이나 넣어주기 위해 ```testAction```이라는 표현을 썼습니다

어떤식으로 표현해야 의도를 더 잘 드러낼 수 있을지 고민이되네요 🤔

hannut91 replied:
테스트가 내가 만든 코드들에 대해서 사용법을 나타낸다고 생각하시면 좋을 것 같아요. 만약 잘못된 액션이 들어왔을 때를 테스트한다면 이렇게 될 것 같아요.

```js
context('with wrong action', () => {
  it('returns previous state', () => {
    const previousState = {};
    const state = reducer(previousState, 'WRONG_ACTION');
    
    expect(state).toEqual(previousState);
  });
});
```

naraekn replied:
wrong action에 대한 test도 넣어야겠습니다 👍
그런데, 제가 원래 의도했던 부분은 ```initialState```으로 세팅하는 부분이기는 했어요ㅎㅎ 그래서 일부로 ```undefined```를 넣어줬습니다!

hannut91 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,26 @@
+    });
+  });
+
+  context('when action type is setCategories', () => {
+    it('gets new categories', () => {
+      const state = reducer(undefined, setCategories(testCategories));
+
+      expect(state.categories).toHaveLength(2);
+    });
+  });
```
testCategories로 업데이트되는 것이라고 의도를 드러내고 싶으면 이렇게 해도 괜찮을 것 같아요

```js
context('when action is setCategories', () => {
  it('updates categories', () => {
    const state = reducer(undefined, setCategories(testCategories));
   
    expect(state.categories).toEqual(testCategories);
  });
});
```

hannut91 reviewed (commented):
잘 진행하고 계세요. 계속 진행해주세요 😀

naraekn committed:
Update test for reducer in case of setCategories

Use toEqual matcher instead of toHaveLength

naraekn committed:
Use fixture in reducer test

Remove testCategories

naraekn committed:
Update store to create store

naraekn committed:
Add mock dispatch to App test

naraekn committed:
Add fetchCategories into services

Use fetchCategories in App

naraekn committed:
Add mock fetchCategories into App test

naraekn committed:
Add thunk and apply middleware

naraekn committed:
Add loadCategories action

naraekn committed:
Refactor App with loadCategories action

naraekn committed:
Add test for CategoriesContainer

naraekn committed:
Implement CategoriesContainer

naraekn committed:
Refactor Categories

Use Categories instead of paragraph
Add useSelector to get categories

naraekn committed:
Add test for Categories

naraekn committed:
Add Categories

naraekn committed:
Refactor App using CategoriesContainer

naraekn committed:
Update test for CategoriesContainer for selectCategory action

Test if selectCategory works

naraekn committed:
Add selectCategory action

naraekn committed:
Add test for reducer in case of selectCategory

naraekn committed:
Update reducer to cover selectCategory action

naraekn committed:
Update CategoriesContainer to add handleClick

naraekn committed:
Update test for Categories

Test if handleClick is called when button is clicked

naraekn committed:
Update Categories to add onClick

naraekn committed:
Fix errors in Categories

Add argument to onClick

hannut91 commented on src/App.jsx:
```diff
@@ -1,7 +1,20 @@
+import { useDispatch } from 'react-redux';
+
+import { loadCategories } from './actions';
+import CategoriesContainer from './CategoriesContainer';
```
액션이랑 컴포넌트는 다르니 여기도 빈 줄을 추가하는게 좋을 것 같아요

hannut91 commented on src/CategoriesContainer.test.jsx:
```diff
@@ -0,0 +1,26 @@
+import CategoriesContainer from './CategoriesContainer';
+
+import { selectCategory } from './actions';
+import { categories } from '../fixture/test-data';
```
여기도 액션과 테스트 데이터는 서로 다르니 빈 줄을 추가하는게 좋을 것 같아요

hannut91 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,35 @@
+import { categories } from '../fixture/test-data';
+
+describe('reducer', () => {
+  const testState = {
+    categories: [],
+  };
```
테스트에 쓰이는 상태라서 `testState`인 것 같아요. `previousState`라고 하는건 어떨까요?

hannut91 reviewed (commented):
테스트도 잘 작성해주셨고 점진적으로 하나씩 개발해나가는 과정을 확인할 수 있어서 좋았습니다. 지금 테스트는 간단한 것만 테스트하고 있는데 시작은 그렇게 간단하고 성공하는 경우를 테스트하는게 맞습니다 ㅎㅎ 먼저 가장 쉬운 경우를 테스트하고(이걸 Happy path라고 부릅니다) 만약 이런 경우는 어떡하지? 예를들어 카테고리 목록을 받아서 보여주는 코드를 작성한 후 만약 여기서 카테고리가 없으면 어떡하지? 라는 고민을 했을 때, 정답이 안떠오르죠. 왜냐하면 정의되어있지 않기 때문이에요. 그러면 나는 카테고리가 비어있으면 비어있다는 메시지가 있었으면 좋겠어라고 결정한다면 테스트를 작성하겠죠

```js
useSelector.mockImplementation((selector) => selector({
  categories: [],
}));

const { getByText } = render(<Categories categories={[]} />);

expect(getByText('카테고리가 없어요')).toBeInTheDocument();
```

그러면 테스트를 실행시키면 실패하겠죠(RED) 그러면 이제 구현하고 다시 테스트를 동작시키면(GREEN) 올바르게 동작하는 걸 테스트가 자동으로 확인해주죠. 이러한 스텝으로 진행하시면 좋을 것 같아요

처음부터 너무 복잡한 것을 테스트하려고 하면 시작하기도 어려울거에요. ㅎㅎ 지금 잘하고 계십니당.

naraekn committed:
Add line breaks

naraekn committed:
Rename testState to previousState

naraekn committed:
Update test for Categories

Test if check sign is rendered when category is selected

naraekn committed:
Add categoryId props to Categories

naraekn committed:
Update Categories to render check sign

naraekn committed:
Refactor Categories test and fix errors

Use describe-context-it
Divide context between selected and non-selected cases

naraekn committed:
Fix fetchCategories url

naraekn committed:
Update test for App

Test if region buttons are rendered

naraekn committed:
Add regions to fixture

naraekn committed:
Implement App with regions

Add regions list with mock data to pass test

naraekn committed:
Add fetchRegions api

naraekn committed:
Add setRegions and loadRegions actions

naraekn committed:
Update reducer test for setRegions action

naraekn committed:
Update and refactor reducer test

Add test for wrong action
Add test code for initial execution
Change to third person singular expressions

naraekn committed:
Update reducer to cover setRegions action

naraekn committed:
Update App to load categories

naraekn committed:
Update regions fixture to match with App test

naraekn committed:
Add selectRegion action

naraekn committed:
Update reducer test for selectRegion action

naraekn committed:
Update reducer to cover selectRegion action

naraekn committed:
Add test for RegionsContainer

naraekn committed:
Add RegionsContainer

naraekn committed:
Add test for Regions

naraekn committed:
Implement Regions

naraekn committed:
Update App to use RegionsContainer

naraekn committed:
Refactor regions location on codes

Change regions to come before categories

naraekn committed:
Modify inappropriate context messages in tests

naraekn commented:
여쭤보려고 했던 부분을 먼저 답해주셨네요ㅎㅎㅎ

사실 저번주에 과제를 할 때는 test-implement-refactor 과정을 최대한 지키기 위해 노력했습니다. (한참 컴포넌트 개발하는 중간에 이전에 작성한 테스트에 오류가 있는걸 발견하면 ```git reset HEAD^```을 해서 다시 테스트부터 작성을 했어요)

그런데 아샬님 강의를 보고, 제가 너무 TDD의 의도를 따르기보다 과정에만 집착했다는 생각이 들었고 (물론 리셋을 너무 많이해서 TDD연습이 제대로 되긴했습니다..) 이번주부터는 개선을 해야겠다고 생각을 했습니다. 

이번주에는 그래서 최대한 리셋을 안하고 진행해보려고 하고 있는데, 너무 막 하는건가(??ㅎㅎㅎ) 싶기도 해서 한 번 여쭤보려고 했었거든요! 멘토님의 피드백을 보고 점진적으로 만들어가되, 어느 정도 기능이 구현 되면 빠진 부분에 대하여 더하고 리팩터링 하는 시간도 가져야겠다는 생각이 드네요😊

hannut91 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,67 @@
+    });
+  });
+
+  context('when action type is setRegions', () => {
+    it('gets new regions', () => {
+      const state = reducer(previousState, setRegions(regions));
+
+      expect(state.regions).toEqual(regions);
+    });
+  });
```
새로운 regions를 얻는 것 보다는 regions으로 갱신된 새로운 상태를 반환하고 있는 거죠? 테스트를 읽어보면

`reducer when action type is setRegions gets new regions`인데 어색한 것 같아요

hannut91 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,67 @@
+    });
+  });
+
+  context('when action type is setCategories', () => {
+    it('gets new categories', () => {
+      const state = reducer(previousState, setCategories(categories));
+
+      expect(state.categories).toEqual(categories);
+    });
+  });
```
여기도 마찬가지로 새로운 카테고리들을 얻는 것 보다는 상태의 카테고리들을 변경한 후 변경된 새로운 상태를 반환하고 있는 거죠

hannut91 reviewed (commented):
거의 다 된 것 같아요 ㅎㅎ 커버리지 결과를 보시고 어떤 부분이 테스트가 덜 되었는지 확인하신 후 테스트를 더 추가해주세요

naraekn committed:
Update test for App with restaurants

naraekn committed:
Implement App with restaurants

Use example data to pass test first

naraekn committed:
Add fetchRestaurants api

naraekn committed:
Add set and load restaraunts action

naraekn committed:
Add test for RestaurantsContainer

naraekn committed:
Implement RestaurantsContainer

naraekn committed:
Add test for Restaurants

naraekn committed:
Add Restaurants

naraekn committed:
Add reducer test for setRestaurants

naraekn committed:
Modify context-it expressions in reducer test

naraekn committed:
Update reducer to cover setRestaurants action

naraekn committed:
Fix errors in RestaurantsContainer

naraekn committed:
Refactor App to use RestaurantsContainer

naraekn committed:
Fix error in RegionsContainer

Modify typo in component name

naraekn committed:
Update test for RegionsContainer

Test if region button occurs loadRestuarants

naraekn committed:
Update RegionsContainer to load restaurants

naraekn committed:
Add parameter to loadRestaurants

naraekn committed:
Update test for CategoriesContainer

Test if dispatch is called twice

naraekn committed:
Update CategoriesContainer to occur loadRestaurants

naraekn committed:
Add conditios to load restaurants

Load restaurants in CategoriesContainer
when regionName is not undefined

Load restaurants in RegionsContainer
when categoryId is not undefined

naraekn committed:
Install redux-mock-store

naraekn committed:
Update test for Categories and RegionsContainer

Add test for regionName is undefined in CategoriesContainer
Add test for categoryId is undefined in RegionsContainer

naraekn commented:
아직 피드백 시간이 아니지만 우선 푸시했습니다.

- e2e 테스트 통과하게 구현했습니다.
- 비동기 액션 부분만 제외하면 테스트 커버리지를 100%로 만들었습니다

슬랙에 올라온 내용보고 비동기 액션 테스트는 시도를 해보겠습니다!
비동기 액션 테스트가 동작하고 나면, 피드백 받은 부분이랑 합쳐서 리팩토링 진행할게요😁

naraekn committed:
Add test for actions

naraekn committed:
Add test for api

naraekn committed:
Update test for api

Test if fetch works

naraekn commented:
```api```랑 ```action``` 테스트를 더하긴 했는데 맞게했는지 모르겠네요ㅠㅠ
이제 테스트 커버리지는 100%로 만들었습니다.

제 노트북에서 ```npm test```를 했을 때는 e2e테스트를 5개 모두 통과했는데, CI에서는 통과를 못하네요 왜 그러는걸까요 흑..😭

이제 남은 시간은,
* CI 통과못하는 원인찾기(?)
* 리팩터링
* (피드백 받고) action, api 테스트 수정
진행하겠습니다!!

hannut91 commented on src/Categories.test.jsx:
```diff
@@ -0,0 +1,53 @@
+import { categories } from '../fixture/test-data';
+
+describe('Categories', () => {
+  const handleClick = jest.fn();
```
지금은 `handleClick`이 다른 테스트에서 사용되고 있지 않지만 만약 공통으로 사용한다면 문제가 될 수도 있어요.

테스트가 실행되기전에 초기화해주면 이를 방지할 수 있습니다.
```js
beforeEach(() => {
  handleClick.mockClear();
});
```

hannut91 commented on src/CategoriesContainer.jsx:
```diff
@@ -0,0 +1,38 @@
+  function handleClick(id) {
+    dispatch(selectCategory(id));
+
+    if (regionName) {
+      dispatch(loadRestaurants({
+        regionName,
+        categoryId: id,
+      }));
+    }
```
카테고리 컨테이너 안에서 region을 참고 하고 있는 것 같아요. `loadRestaruants` 액션 안에서 이 로직이 들어가 있다면 카테고리 컨테이너는 region을 몰라도 잘 동작할거에요

hannut91 commented on src/CategoriesContainer.test.jsx:
```diff
@@ -0,0 +1,58 @@
+
+      useDispatch.mockImplementation(() => dispatch);
+
+      useSelector.mockImplementation((selector) => selector({
+        categories, categoryId, regionName,
+      }));
```
그래서 지금 보면 카테고리 컨테이너 안에서 지역 이름에 의존하고 있잖아요? 이를 분리시키면 좋을 것 같아요

hannut91 commented on src/RegionsContainer.jsx:
```diff
@@ -0,0 +1,38 @@
+  function handleClick(name) {
+    dispatch(selectRegion(name));
+
+    if (categoryId) {
+      dispatch(loadRestaurants({
+        categoryId,
+        regionName: name,
+      }));
+    }
```
여기도 마찬가지로 regions container에서 category를 참조하고 있어요

hannut91 commented on src/reducer.js:
```diff
@@ -0,0 +1,56 @@
+  restaurants: [],
+};
+
+function reducer(state = initialState, action) {
```
`if`나 `switch`없이 이 문제를 해결할 수 있을까요?

hannut91 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,80 @@
+    it('returns initial state', () => {
+      const state = reducer(undefined, 'INITIAL_EXECUTION');
+
+      expect(state.regions).toHaveLength(0);
+      expect(state.categories).toHaveLength(0);
+      expect(state.restaurants).toHaveLength(0);
+
+      expect(state.regionName).toBeUndefined();
+      expect(state.categoryId).toBeUndefined();
```
```js
const initialState = {
  regions: [],
  categories: [],
  restaurants: [],
};

expect(state).toEqual(initialState);
```

이렇게 해주면 초기 상태를 반환한다는 것을 더 드러낼 수 있을 것 같아요

hannut91 commented on src/services/api.test.js:
```diff
@@ -0,0 +1,43 @@
+  fetchRegions, fetchCategories, fetchRestaurants,
+} from './api';
+
+global.fetch = jest.fn(() => Promise.resolve({
+  json: () => Promise.resolve([]),
+}));
```
```js
global.fetch = jest.fn().mockResolvedValue({
  async json() { return [] },
});
```

이렇게 표현할 수도 있습니다

See also
  - https://jestjs.io/docs/en/mock-function-api#mockfnmockresolvedvaluevalue

hannut91 commented on src/services/api.test.js:
```diff
@@ -0,0 +1,43 @@
+    fetch.mockClear();
+  });
+
+  context('when fetchRegions is executed', () => {
+    it('get regions', () => {
+      const data = fetchRegions();
+
+      expect(data).not.toBeNull();
+      expect(fetch).toHaveBeenCalledTimes(1);
+    });
+  });
```
`fetch`가 실행된 것을 확인하는 것 보다는 `fetchRegion`가 무엇을 하는 지 더 드러내면 좋을 것 같아요. 저는 이렇게 했었어요.

```js
describe('fetchRegions', () => {
  global.fetch = jest.fn().mockResolvedValue({
    async json() {
      return REGIONS;
    },
  });

  it('returns regions', async () => {
    const regions = await fetchRegions();

    expect(regions).toEqual(REGIONS);
  });
});
```

naraekn committed:
Clear handleclick before each tests

naraekn committed:
Update test for api to use better expressions

naraekn committed:
Update reducer test for initial execution

naraekn committed:
Remove if statements in reducer

naraekn committed:
Remove unrelated states in Categories and Regions Container

naraekn committed:
Update loadRestaurants actions

naraekn committed:
Update actions test to cover loadRestaurants

naraekn committed:
Add line break to url in fetchRestaurants

naraekn committed:
Add line breaks to actions

naraekn committed:
Remove unnecessary linebreaks in Containers

naraekn commented:
피드백 내용 모두 반영하고, 코드를 한 번 더 정리했습니다.
CI 도 통과를 시켰습니다 😆

hannut91 commented on src/actions.test.js:
```diff
@@ -0,0 +1,64 @@
+    it('get setRegions action', async () => {
+      const store = mockStore({});
+
+      return store.dispatch(loadRegions())
+        .then(() => {
+          const actions = store.getActions();
+          expect(actions[0].type).toBe('setRegions');
+        });
```
`async`, `await`을 쓰시면 더 간단하게 표현할 수 잇을거에요

```js
await store.dispatch(loadRegions());
const actions = store.getActions();

expect(actions[0].type).toBe('setRegions');
```

naraekn replied:
앗 그렇군요ㅠㅠ 다음에는 then 없이 해보겠습니다!

hannut91 reviewed (approved):
고생하셨습니다! 과제 풀이 영상을 통해서 꼭 복습해보세요

hannut91 commented:
이번 과제를 통해서 무엇을 배우셨나요? 배운 것을 다음 과제에도 적용해보고 싶은게 있으신가요?

naraekn commented:
피드백 강의까지 듣고 정리해보겠습니다 정리하고 주간회고에 올리겠습니다 😊
